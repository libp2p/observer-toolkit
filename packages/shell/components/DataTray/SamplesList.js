import React, { useContext, useState } from 'react'
import T from 'prop-types'
import styled from 'styled-components'

import samples from '@libp2p-observer/samples'
import { applySampleData, SetterContext } from '@libp2p-observer/sdk'

const SamplesTray = styled.div`
  position: absolute;
  top: 100%;
  background: ${({ theme }) => theme.color('contrast', 2)};
  margin: 0;
  padding: ${({ theme }) => theme.spacing([0, 2])};
  border-width: ${({ theme }) => theme.spacing(0.5)};
  border-style: solid;
  border-color: ${({ theme }) => theme.color('contrast', 1)};
  border-top-width: 0;
  left: ${({ theme }) => theme.spacing(-2.5)};
  right: ${({ theme }) => theme.spacing(-2.5)};
  border-radius: ${({ theme }) => theme.spacing([0, 0, 2, 2])};
  max-width: 540px;
`

const SampleItem = styled.button`
  border: none;
  cursor: pointer;
  padding: ${({ theme }) => theme.spacing([0.5, 0.5, 0.5, 1])};
  background: ${({ theme }) => theme.color('contrast', 0)};
  ${({ theme }) => theme.text('label', 'medium')}
  color: ${({ theme }) => theme.color('text', 3, 0.7)};
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin: ${({ theme }) => theme.spacing([1, 0])};
  &:active, :focus, :hover {
    background: ${({ theme }) => theme.color('contrast', 0, 0.7)};
  }
`

const SampleImg = styled.img`
  flex-shrink: 1;
  max-width: 75%;
  display: block;
`

function SamplesList({ onLoad }) {
  const [isLoading, setIsLoading] = useState('')
  const { removeData, updateData, updateSource } = useContext(SetterContext)

  function handleClick(samplePath) {
    const handleUploadStart = name => {
      removeData()
      updateSource({ type: 'sample', name })
      setIsLoading(samplePath)
    }
    applySampleData(
      samplePath,
      handleUploadStart,
      handleUploadFinished,
      handleUploadChunk
    )
  }

  function handleUploadFinished() {
    if (isLoading) setIsLoading(false)
    if (onLoad) onLoad()
  }

  function handleUploadChunk(data) {
    updateData(data)
  }

  return (
    <>
      <div>Choose a sample:</div>
      <SamplesTray>
        {samples.map(({ name, img, file }) => {
          const text = (isLoading === file && 'Loading...') || name

          return (
            <SampleItem key={name} onClick={() => handleClick(file)}>
              {text}
              <SampleImg src={img} />
            </SampleItem>
          )
        })}
      </SamplesTray>
    </>
  )
}

SamplesList.propTypes = {
  onLoad: T.func,
}

export default SamplesList
