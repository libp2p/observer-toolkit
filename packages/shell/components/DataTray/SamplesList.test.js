import React from 'react'
import { fireEvent, act } from '@testing-library/react'
import waitForExpect from 'wait-for-expect'
import fs from 'fs'

import {
  renderWithData,
  renderWithShell,
  loadSample,
  within,
} from '@libp2p-observer/testing'
import samples from '@libp2p-observer/samples'

import SamplesList from './SamplesList'

beforeEach(() => {
  window.fetch = async path => {
    const buffer = await fs.promises.readFile(path)
    return {
      ok: true,
      arrayBuffer: async () => {
        const arrayBuffer = new ArrayBuffer(buffer.length)
        const view = new Uint8Array(arrayBuffer)
        let i = 0
        const length = buffer.length
        while (i < length) {
          view[i] = buffer[i]
          i++
        }
        return buffer
      },
    }
  }
})

describe('SamplesList', () => {
  it('renders as expected', () => {
    const { asFragment } = renderWithData(<SamplesList />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('loads sample datasets, updating shell content', async () => {
    const mockFn = jest.fn()

    const { findByText, getByText } = renderWithShell(
      <SamplesList onLoad={mockFn} />
    )
    const { runtime } = loadSample()

    const peerId = runtime.getPeerId()

    const getCurrentPeerId = async () => {
      const shellPeerIdButton = await findByText(/^peer id â€”/i)
      await act(async () => {
        await fireEvent.mouseEnter(shellPeerIdButton)
      })

      const tooltip = await within(shellPeerIdButton).findByRole('tooltip')
      const tooltipText = tooltip.textContent
      const button = within(tooltip).getByRole('button')
      const buttonText = button.textContent
      const currentPeerId = tooltipText.replace(buttonText, '')

      await act(async () => {
        await fireEvent.mouseLeave(shellPeerIdButton)
      })
      return currentPeerId
    }

    expect(await getCurrentPeerId()).toEqual(peerId)

    // Loading a different sample loads a different Peer ID
    const sample2MinButton = getByText('2 minutes')
    await act(async () => {
      await fireEvent.click(sample2MinButton)
      await waitForExpect(() => {
        expect(mockFn).toHaveBeenCalledTimes(1)
      })
    })
    expect(await getCurrentPeerId()).not.toEqual(peerId)

    // Loading the original sample loads the original Peer ID
    const sample1MinButton = getByText('1 minute')
    await act(async () => {
      await fireEvent.click(sample1MinButton)
      await waitForExpect(() => {
        expect(mockFn).toHaveBeenCalledTimes(2)
      })
    })
    expect(await getCurrentPeerId()).toEqual(peerId)
  })
})
