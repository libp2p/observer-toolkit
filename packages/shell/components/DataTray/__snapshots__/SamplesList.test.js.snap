// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SamplesList renders as expected 1`] = `
<DocumentFragment>
  .c0 {
  width: inherit;
  height: inherit;
  display: inherit;
  -webkit-flex-direction: inherit;
  -ms-flex-direction: inherit;
  flex-direction: inherit;
}

.c1 {
  position: absolute;
  top: 100%;
  background: rgb(121,27,137);
  margin: 0;
  padding: 0px 16px;
  border-width: 4px;
  border-style: solid;
  border-color: rgb(78,23,91);
  border-top-width: 0;
  left: -20px;
  right: -20px;
  border-radius: 0px 0px 16px 16px;
  max-width: 540px;
}

.c2 {
  border: none;
  cursor: pointer;
  padding: 4px 4px 4px 8px;
  background: rgb(20,22,45);
  font-family: plex-sans,sans-serif;
  line-height: 1.6em;
  font-size: 9pt;
  line-height: 1.2em;
  color: rgba(255,255,255,0.7);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 8px 0px;
}

.c2:active,
.c2:focus,
.c2:hover {
  background: rgba(20,22,45,0.7);
}

.c3 {
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  max-width: 75%;
  display: block;
}

<div
    class="c0"
  >
    <div>
      Choose a sample:
    </div>
    <div
      class="c1"
    >
      <button
        class="c2"
      >
        1 minute
        <img
          class="c3"
          src="../../packages/samples/samples/sample-1min.png"
        />
      </button>
      <button
        class="c2"
      >
        2 minutes
        <img
          class="c3"
          src="../../packages/samples/samples/sample-2min.png"
        />
      </button>
      <button
        class="c2"
      >
        3 minutes
        <img
          class="c3"
          src="../../packages/samples/samples/sample-3min.png"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
